pkg/graphite/convert/whisperconverter/whisper.go

```diff
@@ -54,53 +54,72 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
 		return nil, fmt.Errorf("whisper file contains no archives for metric: %q", name)
 	}
 
-	totalPoints := 0
-	for _, a := range w.GetArchives() {
-		totalPoints += int(a.Points)
-	}
-
-	// Preallocate space for all allPoints in one slice.
-	allPoints := make([]pointWithPrecision, totalPoints)
-	pIdx := 0
 	// Dump one precision level at a time and write into the output slice.
+	// Its important to remember that the archive with index 0 (first archive)
+	// has the raw data and the highest precision https://graphite.readthedocs.io/en/latest/whisper.html#archives-retention-and-precision
+	seenTs := map[uint32]struct{}{}
+	lastMinTs := uint32(0)
+	var keptPoints []whisper.Point
 	for i, a := range w.GetArchives() {
 		archivePoints, err := w.DumpArchive(i)
 		if err != nil {
 			return nil, fmt.Errorf("failed to dump archive %d from whisper metric %s", i, name)
 		}
-		for j, p := range archivePoints {
-			allPoints[pIdx+j] = pointWithPrecision{p, a.SecondsPerPoint}
+
+		var minArchiveTs, maxArchiveTs uint32
+		// calculate maxArchiveTs
+		for _, p := range archivePoints {
+			// We want to track the max timestamp of the archive because we know
+			// it virtually represents now() and we wont have newer points.
+			// Then the min timestamp of the archive would be maxTs - the archive
+			// retention.
+			if p.Timestamp > maxArchiveTs {
+				maxArchiveTs = p.Timestamp
+			}
 		}
-		pIdx += len(archivePoints)
-	}
 
-	// Points must be in time order.
-	sort.Slice(allPoints, func(i, j int) bool {
-		return allPoints[i].Timestamp < allPoints[j].Timestamp
-	})
+		// calculate minArchiveTs
+		if maxArchiveTs-a.Retention() > maxArchiveTs { // overflow, very big retention
+			minArchiveTs = 0
+		} else {
+			minArchiveTs = maxArchiveTs - a.Retention()
+		}
 
-	trimmedPoints := []whisper.Point{}
-	for i := 0; i < len(allPoints); i++ {
-		// Remove all points of time = 0.
-		if allPoints[i].Timestamp == 0 {
-			continue
+		// For subsequent archives
+		if lastMinTs > 0 {
+			// if we are already in the second or subsequent archive and we had
+			// some points in the prior archives, we want to skip
+			// samples in the previous archives
+			maxArchiveTs = lastMinTs
 		}
-		// There might be duplicate timestamps in different archives. Take the
-		// higher-precision archive value since it's unaggregated.
-		if i > 0 && allPoints[i].Timestamp == allPoints[i-1].Timestamp {
-			if allPoints[i].secondsPerPoint == allPoints[i-1].secondsPerPoint {
-				return nil, fmt.Errorf("duplicate timestamp at same precision in archive %s: %d", name, allPoints[i].Timestamp)
+		lastMinTs = minArchiveTs
+
+		for _, p := range archivePoints {
+			if p.Timestamp < minArchiveTs {
```

Comment:
Shall we check if `minArchiveTs <= p.Timestamp && p.Timestamp <= maxArchiveTs` holds here?

If we do that, we don't need to keep `seenTs`.

Reply:
Actually check for  `minArchiveTs` should be exclusive, ie. `minArchiveTs < p.Timestamp` must hold.

Reply:
Because points are written to all archives at once, all archives will contain points from timestamps that are within the retentions. So we do need a check like this. I think we can do it more efficiently if we pre-sort the points and then do a double-index comparison